src/App/Fossa/ArchiveUploader.hs

```diff
@@ -0,0 +1,140 @@
+  parseJSON = withObject "VendoredDependency" $ \obj ->
+    VendoredDependency <$> obj .: "name"
+      <*> obj .: "path"
+      <*> (unTextLike <$$> obj .:? "version")
+      <* forbidMembers "vendored dependencies" ["type", "license", "url", "description"] obj
+
+uploadArchives :: (Has Diag.Diagnostics sig m, Has (Lift IO) sig m) => ApiOpts -> [VendoredDependency] -> Path Abs Dir -> Path Abs Dir -> m [Archive]
+uploadArchives apiOpts deps arcDir tmpDir = traverse (compressAndUpload apiOpts arcDir tmpDir) deps
+
+compressAndUpload :: (Has Diag.Diagnostics sig m, Has (Lift IO) sig m) => ApiOpts -> Path Abs Dir -> Path Abs Dir -> VendoredDependency -> m Archive
+compressAndUpload apiOpts arcDir tmpDir dependency = do
+  compressedFile <- sendIO $ compressFile tmpDir arcDir (T.unpack $ vendoredPath dependency)
+
+  depVersion <- case vendoredVersion dependency of
+    Nothing -> sendIO $ hashFile compressedFile
+    Just version -> pure version
+
+  signedURL <- Fossa.getSignedURL apiOpts depVersion (vendoredName dependency)
+
+  _ <- Fossa.archiveUpload signedURL compressedFile
+
+  pure $ Archive (vendoredName dependency) depVersion
+
+-- archiveUploadSourceUnit receives a list of vendored dependencies, a root path, and API settings.
+-- Using this information, it uploads each vendored dependency and queues a build for the dependency.
+archiveUploadSourceUnit :: (Has Diag.Diagnostics sig m, Has (Lift IO) sig m) => Path Abs Dir -> ApiOpts -> [VendoredDependency] -> m (Maybe SourceUnit)
+archiveUploadSourceUnit baseDir apiOpts vendoredDeps = do
+  archives <- withSystemTempDir "fossa-temp" (uploadArchives apiOpts vendoredDeps baseDir)
+
+  -- archiveBuildUpload takes archives without Organization information. This orgID is appended when creating the build on the backend.
+  -- We don't care about the response here because if the build has already been queued, we get a 401 response.
+  _ <- Fossa.archiveBuildUpload apiOpts (ArchiveComponents archives)
+
+  -- The organizationID is needed to prefix each locator name. The FOSSA API automatically prefixes the locator when queuing the build
+  -- but not when reading from a source unit.
+  Fossa.Organization orgId _ <- Fossa.getOrganization apiOpts
+  let archivesWithOrganization = updateArcName (T.pack $ show orgId) <$> archives
+
+  pure $ Just $ archivesToSourceUnit archivesWithOrganization
+  where
+    updateArcName :: Text -> Archive -> Archive
+    updateArcName updateText arc = arc{archiveName = updateText <> "/" <> archiveName arc}
+
+-- archiveNoUploadSourceUnit exists for when users run `fossa analyze -o` and do not upload their source units.
+archiveNoUploadSourceUnit :: [VendoredDependency] -> Maybe SourceUnit
+archiveNoUploadSourceUnit deps = Just $ archivesToSourceUnit (unsafeVendoredToArchive <$> deps)
+
+unsafeVendoredToArchive :: VendoredDependency -> Archive
+unsafeVendoredToArchive dep = Archive (vendoredName dep) (fromMaybe "" $ vendoredVersion dep)
```

Comment:
I called it `unsafe` because VendoredDependency has an optional version field whereas Archive requires it. This function is used when `fossa analyze -o` is run to output the deps without having to hash the files to acquire versions.

Reply:
I renamed it to `forceVendoredToArchive` but if you have a better name let me know.